pkg/apis/core/validation/shoot.go

```diff
@@ -1747,29 +1747,48 @@ func validateShootOperation(operation string, shoot *core.Shoot, fldPath *field.
 		return allErrs
 	}
 
-	if !utilfeature.DefaultFeatureGate.Enabled(features.ShootCARotation) {
-		return allErrs
-	}
-
 	switch operation {
 	case v1beta1constants.ShootOperationRotateCAStart:
-		if !isShootReadyForCARotationStart(shoot.Status.LastOperation) {
+		if !utilfeature.DefaultFeatureGate.Enabled(features.ShootCARotation) {
+			return allErrs
+		}
+		if !isShootReadyForRotationStart(shoot.Status.LastOperation) {
 			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot start CA rotation if shoot was not yet created successfully or is not ready for reconciliation"))
 		}
 		if phase := helper.GetShootCARotationPhase(shoot.Status.Credentials); len(phase) > 0 && phase != core.RotationCompleted {
 			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot start CA rotation if .status.credentials.rotation.certificateAuthorities.phase is not 'Completed'"))
 		}
-
 	case v1beta1constants.ShootOperationRotateCAComplete:
+		if !utilfeature.DefaultFeatureGate.Enabled(features.ShootCARotation) {
+			return allErrs
+		}
 		if helper.GetShootCARotationPhase(shoot.Status.Credentials) != core.RotationPrepared {
 			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot complete CA rotation if .status.credentials.rotation.certificateAuthorities.phase is not 'Prepared'"))
 		}
+
+	case v1beta1constants.ShootOperationRotateServiceAccountKeyStart:
+		if !utilfeature.DefaultFeatureGate.Enabled(features.ShootSARotation) {
+			return allErrs
+		}
+		if !isShootReadyForRotationStart(shoot.Status.LastOperation) {
+			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot start service account key rotation if shoot was not yet created successfully or is not ready for reconciliation"))
+		}
+		if phase := helper.GetShootServiceAccountKeyRotationPhase(shoot.Status.Credentials); len(phase) > 0 && phase != core.RotationCompleted {
+			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot start service account key rotation if .status.credentials.rotation.serviceAccountKey.phase is not 'Completed'"))
+		}
+	case v1beta1constants.ShootOperationRotateServiceAccountKeyComplete:
+		if !utilfeature.DefaultFeatureGate.Enabled(features.ShootSARotation) {
+			return allErrs
+		}
+		if helper.GetShootServiceAccountKeyRotationPhase(shoot.Status.Credentials) != core.RotationPrepared {
+			allErrs = append(allErrs, field.Forbidden(fldPath, "cannot complete service account key rotation if .status.credentials.rotation.serviceAccountKey.phase is not 'Prepared'"))
+		}
 	}
 
 	return allErrs
 }
 
-func isShootReadyForCARotationStart(lastOperation *core.LastOperation) bool {
+func isShootReadyForRotationStart(lastOperation *core.LastOperation) bool {
```

Comment:
@acumino raised a good point - do we forbid for example a SA rotation start when the Shoot is already in phase 1 of CA rotation. I assume it is not supported to run both in parallel, right? Do we have a validation about this case?

Reply:
Why would it not be supported?

Reply:
Service account secret has ca.crt. 
So for example when the ca rotation starts and is in phase 1, at this stage will this service account secret contain both ca.crt? Also if at this stage if the user triggers SA rotation and completes it will the new SA token can be used with both CA? Maybe it's very obvious thing that I am not able to get but can you please point to the code which handles these scenarios.

Reply:
> So for example when the ca rotation starts and is in phase 1, at this stage will this service account secret contain both ca.crt?

Yes

> Also if at this stage if the user triggers SA rotation and completes it will the new SA token can be used with both CA? 

CA certificates and SA tokens are not related to each other, so I don't really understand this question.

> Maybe it's very obvious thing that I am not able to get but can you please point to the code which handles these scenarios.

Which code are you looking for/which scenarios?